### PR TITLE
Add more stuff to JSON file

### PR DESCRIPTION
**Pull request by marler8997**
This PR is based on recommendation from @WalterBright (https://github.com/dlang/dmd/pull/7746#issuecomment-359153438).  In order for `rdmd` to support single-invocation, we need a way to get the information it currently gets from "verbose output", but without losing error messages and other stdout messages.  Walter recommended rdmd switch to using the JSON file that dmd generates.  Currently the JSON file is missing most of the information that rdmd requires, so this PR is adding that information to it.

Most of the information needed by rdmd is "semantic data" whereas the JSON file seems to contain "syntax data".  Because of this, I've added the "semantic data" to the end of the JSON file in it's own object called `semantics`.  Here's a list of the new additions and the current status:
* non-root imported modules (ADDED)
* file content imports (i.e. `import("somefile")`) (ADDED)

Along with semantic data, there is some build information that rdmd uses.  I've put this in the beginning of the JSON file in an object named "buildInfo". It contains the following:
* target library file if `-lib` is used (ADDED)
* the compiler config file (ADDED)
* the compiler binary file itself (ADDED)

### CODE REVIEW

**Comment by marler8997** [comment]
Hmmm, this is going to make testing more difficult. Now that the JSON file contains semantic data, it's contents are different for each platform, and also changes depending on the order that files were passed to the compiler.

**Comment by timotheecour** [comment]
1 quick sanity check: seeing whether on a large project, adding  -Xf=file
slows down compilation (or consumes more memory ?) than without. Because
each function declaeation (+other things) gets serialized. If so, we'd need
a syntax like -X=imports,versions,libs to control what gets exported.

in any case, its a less gacky version than -v

**Comment by marler8997** [comment]
Based on the code, I don't anticipate json code generation having a big performance impact.  However, if it does then filtering can always be added later as an after thought.  Maybe:
```
-X (enable json code generation)
-Xf=<file> (set file to output json to)
-Xo=-functions  (An example interface to filter out function from json code generation)
```

**Comment by CyberShadow** [comment]
> `-Xo=-functions  (An example interface to filter out function from json code generation)`

BTW, if we ever do need to add that, it should probably be a list of things to include rather than a list of things to exclude. Tools that would make use of this option will know exactly what they want, so any information that's added later will not be useful to them.

**Comment by marler8997** [comment]
yeah good point

**Comment by marler8997** [comment]
@WalterBright We may want to try to give this PR some attention early on because if this PR along with the changes to rdmd don't get in before the next release, then we will have an extra use case to handle, namely, the case where you have a compiler that supports `-i` but doesn't generate the JSON output that is needed.  I'd like to avoid that if possible.

File: src/dmd/json.d
Comment (timotheecour) [inline src/dmd/json.d]: how about:
property("name", m.md ? m.md.toChars() : "");
to make each entry have each field?

File: src/dmd/json.d
Comment (timotheecour) [inline src/dmd/json.d]: how about `absolutePath ( m.srcfile.toChars() )` so that if rdmd is called from a different directory, it'll still work?

File: src/dmd/json.d
Comment (marler8997) [reply src/dmd/json.d]: I was following the convention established here (https://github.com/dlang/dmd/pull/7757/files/a90c983d6e436e46dde3b1ee07fdd280ea1396e7#diff-1d73c979f233c43e47fb88fac32c21b3L475)

I don't really see much difference whether or not there is an empty name or no name at all, but following the existing convention seems the better choice.

File: src/dmd/json.d
Comment (marler8997) [reply src/dmd/json.d]: hmmmm, interesting thought.  I'd have to think about that one.  This is mimicking the current functionality from the verbose output, but maybe it can be improved.  I'll come back to this.

Comment (timotheecour) [reply src/dmd/json.d]: NOTE: "contentImports" appears to use absolutePath (i tried with -J. and `import("foo.txt")` ), so I would make sense to be consistent. I suggest using absolutePath for `file` as well (and wherever else it makes sense)

Comment (marler8997) [reply src/dmd/json.d]: I believe that would be changing behavior though...we could end up with a regression if we change the behavior.

Comment (timotheecour) [reply src/dmd/json.d]: well, that info wasn't in json output before... so it's one of those gray area things.
one option is to allow configuring json output (as I suggested somewhere else), which is OK if we have good way to customize, eg:

`-Xf=file -Xopt=abspaths,-functions`

Comment (marler8997) [reply src/dmd/json.d]: What I mean is that `rdmd` will be using that field the same way it was using the `import   ` line from the verbose output.  And in that case it was using the same `m.srcfile.toChars()` expression.  If we use the absolute path instead moving forward, I just want to make sure it won't break anything.

Comment (marler8997) [reply src/dmd/json.d]: Now that I think about it, there might be a bug in the current design.  If you build with rdmd, and then you change directories and build again, then all the relative module filenames will be incorrect.  It might just force you to do a rebuild, but that's no good.  We could fix this using absolute paths, or we could save off the original CWD that rdmd was using when it first performed the build.

File: src/dmd/json.d
Comment (timotheecour) [inline src/dmd/json.d]: BUG! missing separaters in generated JSON
"contentImports" : [
"/tmp/myfile.txt""/tmp/myfile.txt"
    ]

File: src/dmd/json.d
Comment (CyberShadow) [reply src/dmd/json.d]: There should be a test for that in the rdmd test suite, FWIW.

Comment (timotheecour) [reply src/dmd/json.d]: there may even be corner cases where it picks up wrong file eg:

ls
main.d
fun.d
temp/fun.d // slightly different version main.d

rdmd main.d
cd temp
rdmd ../main.d
=> can it get confused and get the wrong fun.d?

Here's a good compromise:

add a CWD field, and when rdmd reads a path (any path) call  buildPath(json.CWD, json.path) (does the right thing w absolute paths)

and allow for:
rdmd main.d
cd temp
rdmd ../main.d => should reuse cache and not rebuild

Comment (timotheecour) [reply src/dmd/json.d]: but we should definitely make behavior consistent between contentImports and modules.file:
import("fun.txt") should not get auto-transformed in an absolute path either.

Comment (marler8997) [reply src/dmd/json.d]: I'm thinking that the JSON file saves the CWD in the new `buildInfo` object.

**Review by wilzbach — COMMENTED** [review]
Do we really need to generate the JSON file?
It makes future updates of the JSON file more difficult as the line numbers on the diff won't reflect the ones in the generated script. Also it's easy to screw up the JSON generation. How about moving the things that are platform-specific into a different test?

File: test/compilable/extra-files/gen_json_out.d
Comment (marler8997) [reply test/compilable/extra-files/gen_json_out.d]: they are "grepped" out along with some other types of lines (see json-postscript.sh).  I thought keeping an example in the actual expected results didn't hurt though.

**Comment by wilzbach** [comment]
>> Do we really need to generate the JSON file?
>> It makes future updates of the JSON file more difficult as the line numbers on the diff won't reflect the ones in the generated script. Also it's easy to screw up the JSON generation. How about moving the things that are platform-specific into a different test?
> Could you elaborate on your idea a bit more?

Here's what I would do:
- Filter out all dynamic content from the main `json.d` test - I think you already realized that `compilerBinary` and `version` are rather tricky to handle
- Create a special test for tests, which are a bit more error tolerant, e.g. by using `sed` to replace the version to a static string

> I'm thinking a better idea might be to create a small json comparer program. Since we are adding a list of all the modules that are being imported (directly or indirectly) this means that dmd and druntime/phobos will have to have a knife edge change whenever that import structure changes

I would keep it simply and do checks like:
- `object` is part of the list (the path might be different for LDC/GDC)
- the number of modules is > 0
- maybe a sanity check with `-v`

**Comment by marler8997** [comment]
I'm thinking we create a "sanitize" program that takes a JSON file, and sanitizes it removing all the environment specific stuff.  Then we can remove the JSON generation code ("gen_json_out.d") and bring back the raw `json.out`.  Then just "sanitize" the JSON generated by the test.  The sanitizer will need to remove all the druntime/phobos modules from the semantic module list.

**Comment by timotheecour** [comment]
the json file comparison should be done by parsing the json, then filtering out nodes you want to compare, seems more robust than grep / sed based tools

**Comment by marler8997** [comment]
I've created a `sanitize_json` tool that takes care of removing all the environment specific stuff.  Note that I included a json parser in the test suite because I needed the parser to maintain the original order of the object members.   This allowed me to bring back the original `json.out` and we can still use `diff` to compare the expected vs actual json output.

File: src/dmd/json.d
Comment (timotheecour) [inline src/dmd/json.d]: is that a concern?
https://msdn.microsoft.com/en-us/library/ms235450.aspx
This POSIX function is deprecated. Use the ISO C++ conformant _getcwd instead.

Comment (marler8997) [reply src/dmd/json.d]: This declaration was copied from `dmodule.d`.  It it's a problem then the fix should be a separate PR.

Comment (wilzbach) [reply src/dmd/json.d]: Always open an issue about such stuff, otherwise it well get forgotten. 
I just did so for this: https://issues.dlang.org/show_bug.cgi?id=18291

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Official code should use `shared` whenever possible, though in this case neither is needed.

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Why not use `stdin` if no arguments are provided?
If you don't like this, "-" is a common choice. If `-Xf` would support using stdout as output, this would allow to completely avoid the need of a temporary file.

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: `ref`

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: `ref`

Comment (wilzbach) [reply test/sanitize_json.d]: You could use templates here, e.g. `remove!string`

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: `startsWith` is variadic and supports multiple needles.

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Without the special casing for Windows:

```d
modulesArray
.map!(m => object)
.filter!(m => !m.getAs!string("name").startsWith("std.", "core.", "etc."))
.each!(m => newModules.put(JsonValue(m)));
```

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Maybe `JsonWriter`

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: check out `assumeUnique` from `std.exception`.

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Why is this necessary. You don't use any C function?

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Not used.

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Not used

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Only `appender` is used. It's generally a best practice to do local imports - that way it's easy not to forget to update them when code gets modified / refactored.

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Not used.

File: test/json.d
Comment (wilzbach) [inline test/json.d]: Json only supports floating point numbers:

> A number is a sequence of decimal digits with no superfluous leading zero. It may have a preceding minus
sign (U+002D). It may have a fractional part prefixed by a decimal point (U+002E). It may have an exponent,
prefixed by e (U+0065) or E (U+0045) and optionally + (U+002B) or – (U+002D). The digits are the code
points U+0030 through U+0039.

http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf

File: test/json.d
Comment (wilzbach) [inline test/json.d]: Allman braces

File: test/json.d
Comment (wilzbach) [inline test/json.d]: How about: `text.filter!(a == '\n').walkLength`

Also there is `newline` in `std.ascii` and `std.uni`

File: test/json.d
Comment (wilzbach) [inline test/json.d]: Why not use `std.ascii`?

**Review by wilzbach — COMMENTED** [review]
I think we should try to avoid re-doing a custom Json parser if possible. Are you sure simply CLI tools don't work?

**Comment by marler8997** [comment]
> I think we should try to avoid re-doing a custom Json parser if possible. Are you sure simply CLI tools don't work?

@timotheecour and I discussed the best solution here.

The first question was whether we should compare the JSON via data structures or at the file level.  I've concluded that performing the final diff on the file level is the most helpful.  Since the `json.out` file is what needs to be maintained, then showing exactly where there are differences in the file itself is more useful than trying to showing a diff between 2 JSON data structures.

Given this, we need a way to sanitize a JSON file (remove/normalize environment specific stuff).  Up to this point we could do this by simply removing the lines that contained environment specific stuff, however, with the newly added data, we need to remove objects that can span multiple lines. Given this, moving to a sanitizer that operates on JSON data structures makes sense here.  However, when we load in the JSON write it back out, we want to preserve the original order for object members.  This is a problem since all the JSON parsers that I know of don't preserve this order, hence the need for a new JSON parser.  Note that I have made an attempt to add support for this in `std.json` (https://github.com/dlang/phobos/pull/6059), if that gets accepted we can swap in that implementation and remove this one.

Also note that by using a small D program to sanitize the JSON, we can keep more of the information in the file to compare.  For example, we are now keeping the module filenames by normalizing the directory separators to `/`, whereas before we were completely removing them.  Since we are adding new JSON, having the sanitation logic in D also makes it easier to maintain moving forward.

I don't think adding a JSON parser is ideal but here are some thoughts that help justify it
1. JSON is very simple, adding a parser for a specific purpose is quite easy.  I wrote this one from scratch in less than an hour
2. JSON is old now and is not going to change.
3. This JSON parser only has to support the JSON that dmd outputs, which is subset of the already small JSON spec.

File: test/sanitize_json.d
Comment (marler8997) [reply test/sanitize_json.d]: with stdin you can't use the file size to allocate the buffer to read the string into.   also by reading in a file you can use that filename in error messages :)

File: test/sanitize_json.d
Comment (marler8997) [reply test/sanitize_json.d]: The JSON parser relies on the file content being null-terminated.  Makes the parser simpler.

File: test/sanitize_json.d
Comment (marler8997) [reply test/sanitize_json.d]: My brain doesn't think in these algorithm functions very well :(  Is this easier for you to understand than the current loop?

File: test/sanitize_json.d
Comment (marler8997) [reply test/sanitize_json.d]: This is just personal preference, but when a function is modifying a value that's passed into it, I prefer to see some indication of this when calling the function, i.e.
```
// I prefer
removeString(&myValue);
// rather than
removestring(myValue);
```
To me this make it obvious that the function is modifying the value.  Note that I only use this convention for value types.  For example if you had a function that modified an integer as a parameter such as `square`, I would rather have `void square(int* i)` rather than `void square(ref int i)` because I would prefer to see `square(&myInt)` rather than `square(myInt)`.

Comment (wilzbach) [reply test/sanitize_json.d]: How about `out` then?

See also: https://github.com/dlang/dmd/pull/7754/files

Comment (marler8997) [reply test/sanitize_json.d]: That has the same problem doesn't it?  You can't see from the call that the value is being modified.
```D
void squareUsingOut(out int i);
// vs
void squareUsingPtr(int* i);

int i = 100;
squareUsingOut(i); // not obvious that i is being modified
squareUsingPtr(&i); // obvious that square could be modifying i
```

**Comment by CyberShadow** [comment]
Was the option to add a secret switch to DMD to only emit JSON fields that are worth testing considered?

**Comment by marler8997** [comment]
Hmmmm, I'm sure the argument against this would be that this adds logic to the compiler that doesn't add any value to the compiler itself.  It just works around an issue with the test suite.

**Comment by timotheecour** [comment]
> Was the option to add a secret switch to DMD to only emit JSON fields that are worth testing considered?

plus there's issue of certain fields like `contentImports` that always output using `absolutePath` (which IMO is a design flaw but that's another topic), so that field would have to be skipped from testing (ie testing won't catch regressions related to `contentImports`)

**Comment by timotheecour** [comment]
* top-level msg: `s/target library file is -lib is used /target library file if -lib is used /`
* how about adding `-I` (import search paths) to the json?

**Comment by marler8997** [comment]
> * how about adding -I (import search paths) to the json?

Yeah I could see that being useful for sure. I'll add em.

**Comment by marler8997** [comment]
@andralex has determined that adding policies to `std.json` is not a desirable enhancement (https://github.com/dlang/phobos/pull/6059#issuecomment-361430810).  Given that we can't use `std.json` to sanitize the file, using the provided JSON parser seems like a solid option to me.

**Comment by CyberShadow** [comment]
> Given that now we can't use std.json to sanitize the file,

Sorry, I don't think we finished [the previous conversation](https://github.com/dlang/phobos/pull/6084#issuecomment-361068225). Why not use `std.json`, and rely on its current behavior of sorting key names lexicographically? Won't that allow achieving the same goal without adding a JSON parser to the DMD repository?

**Comment by marler8997** [comment]
That would work "functionally".  The only drawback is that the two files won't be in the same order that was generated from DMD, which would look odd.  The order of the fields as generated by DMD makes sense, i.e. name first, followed by other identifying properties, followed by more verbose fields such as large objects.  Would be nice to keep it in the same order.  I refuse to accept that sanitizing a JSON file while preserving the original object field order is an insurmountable task in D :)

**Comment by CyberShadow** [comment]
> The only drawback is that the two files won't be in the same order that was generated from DMD, which would look odd.

IMHO that's fine, considering the circumstances, but it's not up to me to decide.

> I refuse to accept that sanitizing a JSON file while preserving the original object field order is an insurmountable task in D :)

Certainly not insurmountable, but it may be worth keeping our feet on the ground in this case. It's just a test case file that hopefully won't be edited / looked at often at all, so maybe just go for the simplest solution.

**Comment by marler8997** [comment]
> Certainly not insurmountable, but it may be worth keeping our feet on the ground in this case. It's just a test case file that hopefully won't be edited / looked at often at all, so maybe just go for the simplest solution.

Yes I see your point.  I'll concede this point and go with your solution.  Be prepared for a large `json.out` diff :)

**Comment by marler8997** [comment]
Ok I caved, the JSON sanitizer uses `std.json` so all the object fields are sorted alphabetically.

**Comment by marler8997** [comment]
Ping.  Trying to get this PR finalized and integrated so we have some time before the 2.079 release to modify rdmd to use it (https://github.com/dlang/tools/pull/292).

**Comment by JinShil** [comment]
Restarted Jenkins

File: test/sanitize_json.d
Comment (wilzbach) [inline test/sanitize_json.d]: Not a fan of this (I know that's how the status quo is).
Ideally this would be configurable via a flag, s.t. special platform-specific, short tests which test that the `deco` is correctly set in the json output can be added, but that's definitely not something which shouldn't block this PR.

**Review by wilzbach — APPROVED** [review]
Looks good, but please add a short changelog entry.

File: test/sanitize_json.d
Comment (marler8997) [reply test/sanitize_json.d]: Ok, I added a `--keep-deco` flag that can be used by tests in the future.

**Comment by marler8997** [comment]
Added the changelog

File: src/dmd/json.d
Comment (wilzbach) [inline src/dmd/json.d]: This isn't covered.

**Review by wilzbach — APPROVED** [review]
Thanks a lot for putting all the hard work into this!
I'm going to merge this now to unblock your work on rdmd - could you do a quick follow up to cover the uncovered output of the imported files? Thanks!

File: test/sanitize_json.d
Comment (timotheecour) [reply test/sanitize_json.d]: > with stdin you can't use the file size to allocate the buffer to read the string into
just consume all stdin and that solves this problem; doesn't need to be processing input line by line for this use case.

> also by reading in a file you can use that filename in error messages :)
```
error: at __stdin.d:(12,10) { "foo": bar"} expected `"` got `b`
```

NOTE: __stdin.d is what is used in src/dmd/mars.d for same use case:
```
 if (strcmp(arguments[i + 1], "-") == 0)
*                        files.push("__stdin.d");
```

File: test/sanitize_json.d
Comment (timotheecour) [inline test/sanitize_json.d]: stderr?

**Pull request merged**
